tools/unit-tests/unit-extflash.c

```diff
@@ -0,0 +1,287 @@
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * wolfBoot is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1335, USA
+ */
+
+/* Option to enable sign tool debugging */
+/* Must also define DEBUG_WOLFSSL in user_settings.h */
+#define FLASH_SIZE (33 * 1024)
+#define WOLFBOOT_HASH_SHA256
+#define IMAGE_HEADER_SIZE 256
+#define EXT_ENCRYPTED
+#define EXT_FLASH
+#define ENCRYPT_WITH_AES256
+
+#define WOLFSSL_AES_COUNTER 
+#define WOLFSSL_AES_DIRECT
+#define WOLFSSL_AES_256
+#define WC_NO_HARDEN
+
+#define WOLFSSL_USER_SETTINGS
+#define ENCRYPT_KEY "123456789abcdef0123456789abcdef0123456789abcdef"
+#define UNIT_TEST
+#include <stdio.h>
+#include <check.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <string.h>
+#include "user_settings.h"
+
+#include "wolfcrypt/src/aes.c"
+#include "libwolfboot.c"
+
+/* Mocks */
+
+static int locked = 0;
+
+void hal_init(void)
+{
+}
+int hal_flash_write(uint32_t address, const uint8_t *data, int len)
+{
+    return 0;
+}
+int hal_flash_erase(uint32_t address, int len)
+{
+    return 0;
+}
+void hal_flash_unlock(void)
+{
+    //fail_unless(locked, "Double unlock detected\n");
+    locked--;
+}
+void hal_flash_lock(void)
+{
+    //fail_if(locked, "Double lock detected\n");
+    locked++;
+}
+
+void hal_prepare_boot(void)
+{
+}
+
+/* Emulation of external flash with a static buffer of 32KB (update) + 1KB (swap) */
+uint8_t flash[FLASH_SIZE];
+
+/* Mocks for ext_flash_read, ext_flash_write, and ext_flash_erase functions */
+int ext_flash_read(uintptr_t address, uint8_t *data, int len) {
+    printf("Called ext_flash_read %p %p %d\n", address, data, len);
+
+    /* Check that the read address and size are within the bounds of the flash memory */
+    ck_assert_int_le(address + len, FLASH_SIZE);
+
+    /* Copy the data from the flash memory to the output buffer */
+    memcpy(data, &flash[address], len);
+
+    return len;
+}
+
+int ext_flash_write(uintptr_t address, const uint8_t *data, int len) {
+    printf("Called ext_flash_write %p %p %d\n", address, data, len);
+    /* Check that the write address and size are within the bounds of the flash memory */
+    ck_assert_int_le(address + len, FLASH_SIZE);
+
+    /* Copy the data from the input buffer to the flash memory */
+    memcpy(&flash[address], data, len);
+
+    return 0;
+}
+
+int ext_flash_erase(uintptr_t address, int len) {
+    /* Check that the erase address and size are within the bounds of the flash memory */
+    ck_assert_int_le(address + len, FLASH_SIZE);
+
```

Comment:
does it make sense to add a check that both address and len are aligned to WOLFBOOT_SECTOR_SIZE?
otherwise the memset below should be modified to also erase the last partial block